KOGITO-452 Domain Explorer Version-1

Sara4994 opened the pull request.

cristianonicolai commented on packages/management-console/src/components/Organisms/DomainExplorerColumnPicker/DomainExplorerColumnPicker.tsx:
```diff
@@ -0,0 +1,239 @@
+  const childSelectionItems = (_data, title, ...attr) => {
+    let nestedTitles = '';
+    b = _data.map(group => {
+      const label = title + '>' + attr.join();
```
```suggestion
      const label = title + ' > ' + attr.join();
```
just adding a extra space here, before and after the string.

Sara4994 replied:
done

cristianonicolai commented on packages/management-console/src/components/Organisms/DomainExplorerColumnPicker/DomainExplorerColumnPicker.tsx:
```diff
@@ -0,0 +1,239 @@
+    try {
+      await axios
+        .post(
+          'http://localhost:4000/graphql',
```
should we create a new query type for this? That would probably avoid having the data index URL in here as well.

Sara4994 commented on packages/management-console/src/components/Templates/DashboardComponent/Dashboard.tsx:
```diff
@@ -1,24 +1,69 @@
+} from '@patternfly/react-core';
+import React, { useState } from 'react';
+import { Redirect, Route, Link } from 'react-router-dom';
+// import HeaderComponent from '../../Organisms/PageHeaderComponent/HeaderComponent';
```
done

Sara4994 commented on packages/management-console/src/components/Organisms/DomainExplorerTable/DomainExplorerTable.tsx:
```diff
@@ -0,0 +1,91 @@
+          onSelect={onRowSelect}
+          aria-label="Filterable Table Demo"
+        >
+          <TableHeader />
```
removed

Sara4994 commented on packages/management-console/src/components/Organisms/DomainExplorerTable/DomainExplorerTable.tsx:
```diff
@@ -0,0 +1,91 @@
+          cells={finalKeys}
+          rows={values}
+          onSelect={onRowSelect}
+          aria-label="Filterable Table Demo"
```
done

cristianonicolai commented:
@Sara4994 we also need to include the page header and breadcrumb for the Domain Explorer page, similar to the one in the Process Instances.

![header](https://user-images.githubusercontent.com/570894/74400588-12f88c00-4e6a-11ea-9cbc-598994afb5a5.png)

Sara4994 commented:
@cristianonicolai All the above comments are taken care and PR is updated with latest code.

cristianonicolai commented:
@Sara4994 overall looks good but I'm getting some weird results when switching domain and columns, see screenshot:

![domain](https://user-images.githubusercontent.com/570894/74424331-75b74b00-4e9d-11ea-8c4d-2e907366dd18.png)

Sara4994 commented:
@cristianonicolai Can you please let me know the selections that you have made. Because i am getting such result for same selection(lastUpdated)
![Screenshot from 2020-02-13 16-06-19](https://user-images.githubusercontent.com/40579476/74426183-f44ec100-4e7a-11ea-8f98-974da541f02b.png)

Sara4994 commented:
![Screenshot from 2020-02-13 16-06-19](https://user-images.githubusercontent.com/40579476/74426224-003a8300-4e7b-11ea-836b-34d46b257f7a.png)

Sara4994 commented:
> @Sara4994 overall looks good but I'm getting some weird results when switching domain and columns, see screenshot:
> 
> ![domain](https://user-images.githubusercontent.com/570894/74424331-75b74b00-4e9d-11ea-8c4d-2e907366dd18.png)
> 
> It seems that some state is kept when you change domains, if the user does that, we should clear the table completely.

@cristianonicolai fixed this bug as well.

srambach commented:
The width of some items are being constrained by the grid and then things start overlapping and wrapping out of their containers. We really don't need to be using the grid layout for this at all, either for the toolbar or for the content section of the page. The DataToolbar should do all the correct spacing.
We talked about moving the column selector to the left, so if we do that you also don't need to worry about how to get it to right align. And can we move the button to the right of the selector?

And as you can see in the Filterable Table demo, just putting the DataToolbar and then the Table, with no grid, is sufficient. https://www.patternfly.org/v4/documentation/react/demos/filtertabledemo

srambach commented:
Is the empty state part of this PR? (probably better as another since this one has been a lot?)

Sara4994 commented:
@srambach I have just removed grid stuffs and added DataToolBar components, also moved column picker to left as mentioned. Added Empty state as well.. Please have a look if it looks fine to you...

cristianonicolai commented on packages/management-console/src/components/Organisms/DomainExplorerColumnPicker/DomainExplorerColumnPicker.tsx:
```diff
@@ -0,0 +1,264 @@
+    if (columnPickerType) {
+      try {
+        await axios
+          .post(
```
review if apollo client can be used instead.

cristianonicolai commented:
@Sara4994 when we have a domain with multiple attributes in the root level, there is quite a big space between then, see:

![image](https://user-images.githubusercontent.com/570894/74511007-b15f1d00-4f50-11ea-97b3-b16c6a129dd7.png)

Sara4994 commented on packages/management-console/src/graphql/queries.tsx:
```diff
@@ -78,3 +78,109 @@ const GET_PROCESS_INSTANCE = gql`
 `;
+
+const GET_PICKER = gql`
+  query getPicker($columnPickerType: String!) {
```
done

Sara4994 commented on packages/management-console/src/components/Organisms/DomainExplorerTable/DomainExplorerTable.tsx:
```diff
@@ -0,0 +1,103 @@
+          <EmptyState>
+            <EmptyStateIcon icon={SearchIcon} />
+            <Title headingLevel="h5" size="lg">
+              Nothing to display
```
done

Sara4994 replied:
done

Sara4994 commented:
> @Sara4994 when we have a domain with multiple attributes in the root level, there is quite a big space between then, see:
> 
> ![image](https://user-images.githubusercontent.com/570894/74511007-b15f1d00-4f50-11ea-97b3-b16c6a129dd7.png)



> @Sara4994 when we have a domain with multiple attributes in the root level, there is quite a big space between then, see:
> 
> ![image](https://user-images.githubusercontent.com/570894/74511007-b15f1d00-4f50-11ea-97b3-b16c6a129dd7.png)

@cristianonicolai ofcourse yes... because we have just given empty string for SelectGroup.. We dont have other choice right now in PF. Even @srambach suggested me same solution, to provide empty string.

Sara4994 commented on packages/management-console/src/graphql/queries.tsx:
```diff
@@ -78,3 +78,109 @@ const GET_PROCESS_INSTANCE = gql`
+`;
+
+const GET_TYPES = gql`
+  query getTypes($currentArgument: String!) {
```
done

Sara4994 commented on packages/management-console/src/graphql/queries.tsx:
```diff
@@ -78,3 +78,109 @@ const GET_PROCESS_INSTANCE = gql`
+`;
+
+const GET_SCHEMA = gql`
+  query getSchema($currentQuery: String!) {
```
done

Sara4994 replied:
done

Sara4994 commented on packages/management-console/src/components/Organisms/DomainExplorerColumnPicker/DomainExplorerColumnPicker.tsx:
```diff
@@ -0,0 +1,264 @@
+    );
+  };
+
+  let b;
```
done

cristianonicolai commented:
@Sara4994 sure, but can't have a single SelectGroup with empty string? It seems that we're creating one for enry single property.

Sara4994 commented on packages/management-console/src/graphql/queries.tsx:
```diff
@@ -78,3 +78,109 @@ const GET_PROCESS_INSTANCE = gql`
+`;
+
+const GET_QUERY = gql`
+  query getQuery {
```
done

Sara4994 commented on packages/management-console/src/graphql/queries.tsx:
```diff
@@ -78,3 +78,109 @@ const GET_PROCESS_INSTANCE = gql`
+`;
+
+const GET_TYPES = gql`
+  query getTypes($currentArgument: String!) {
```
done

srambach commented:
> @Sara4994 sure, but can't have a single SelectGroup with empty string? It seems that we're creating one for enry single property.

Exactly what I was going to suggest. That should provide good spacing.

srambach commented:
I see a few things:
- When switching to a different domain and back again, would it be possible to remember what you had selected previously for that domain?
- When too many columns are selected, the table overflows and there is no horizontal scroll. I'll look at where we want to put the overflow. We might want to wrap the column in one more div for that.
- After adding many columns and the removing some, I see columns with no headers, and they are not getting enough space.
![image](https://user-images.githubusercontent.com/19825616/74549984-970c4b80-4f1e-11ea-8fcc-04d902ab50eb.png)
- Here I had removed bookingNumber but it and another field still appeared (without a column header)
![image](https://user-images.githubusercontent.com/19825616/74550233-08e49500-4f1f-11ea-9ff1-8cb977a51b24.png)
- I selected a few fields, applied it, and then unchecked all of them. "Apply Columns" doesn't remove them to go back to the empty state.
- I'm seeing the removed column data left over even when the column is removed sometimes. Here I added and then removed columns. Also, there used to be 3 rows and suddenly there are 4. It's hard to tell what's going on since most of the fields are identical.
![image](https://user-images.githubusercontent.com/19825616/74551359-08e59480-4f21-11ea-9c15-71cfda8c5b35.png)
Sorry, I know this is a lot for one comment. Let me know if you prefer that I put each thing in a different comment going forward.

srambach commented:
One more - On Travels, there are some shared names between `traveller` and `visaApplication` and I'm seeing the checkbox check both at the same time.

srambach commented:
Another thought - @kywalker-rh and I were talking about using `/` instead of `>` to separate the attributes. So `flight / arrival`? @cristianonicolai what do you think? Also, with or without spaces? I'm thinking with spaces to make it easier to read.

cristianonicolai commented:
@srambach agree to use `/` with space.

@srambach regarding your point for remembering the selection across domains, I think we can postpone that, one because we will introduce a landing page in between any domain change and second that we also should look into how to save searches/filters and columns, similar to v7 approach. Apart from that, it all seems reasonable issues that we need to look into.

Sara4994 commented:
> @Sara4994 when we have a domain with multiple attributes in the root level, there is quite a big space between then, see:
> 
> ![image](https://user-images.githubusercontent.com/570894/74511007-b15f1d00-4f50-11ea-97b3-b16c6a129dd7.png)

This is fixed

Sara4994 commented:
> Another thought - @kywalker-rh and I were talking about using `/` instead of `>` to separate the attributes. So `flight / arrival`? @cristianonicolai what do you think? Also, with or without spaces? I'm thinking with spaces to make it easier to read.

done

Sara4994 commented:
> One more - On Travels, there are some shared names between `traveller` and `visaApplication` and I'm seeing the checkbox check both at the same time.

Fixed

srambach commented:
@Sara4994 When you uncheck a column, the list changes immediately to the empty state until the button is pressed. Checking an additional column doesn't do this. Can we have the list stay in place until the user clicks "Apply columns" again?

srambach commented:
@Sara4994 At the narrower screen width where the filters collapse, clicking on the filter goes to a blank screen.

![SelectDomain-emptyscreen](https://user-images.githubusercontent.com/19825616/74662041-286dfe80-5167-11ea-8f65-0c0560ae451c.gif)

srambach commented on packages/management-console/src/components/Organisms/DomainExplorerTable/DomainExplorerTable.tsx:
```diff
@@ -0,0 +1,112 @@
+              Select a Domain from dropdown to see content
+            </EmptyStateBody>
+          </EmptyState>
+        </Bullseye>
```
I suggest adding a Card component around the Bullseye, which will make the white background to match the table. Also, I suggest some wording changes:
`        <Card component={'div'}>
          <CardBody>
            <Bullseye>
              <EmptyState>
                <EmptyStateIcon icon={SearchIcon} />
                <Title headingLevel="h5" size="lg">
                  No domain data to display
                </Title>
                <EmptyStateBody>
                  To see content, select a domain from the dropdown, and then select the columns to view.
                </EmptyStateBody>
              </EmptyState>
            </Bullseye>
          </CardBody>
        </Card>
`

Sara4994 replied:
done

srambach commented on packages/management-console/src/components/Organisms/DomainExplorerTable/DomainExplorerTable.tsx:
```diff
@@ -0,0 +1,112 @@
+  EmptyState,
+  EmptyStateIcon,
+  EmptyStateBody,
+  EmptyStateSecondaryActions,
```
I think this is unused, so it can be removed

Sara4994 replied:
Removed EmptyStateSecondaryActions, other things are used

srambach commented on packages/management-console/src/components/Organisms/DomainExplorerTable/DomainExplorerTable.tsx:
```diff
@@ -0,0 +1,112 @@
+  EmptyStateBody,
+  EmptyStateSecondaryActions,
+  Bullseye,
+  Button
```
This also looks like it's not used here, so it can be removed

Sara4994 replied:
Remove Button and EmptyStateSecondaryActions... Other things are used

srambach commented on packages/management-console/src/components/Templates/DomainExplorerPage/DomainExplorerPage.css:
```diff
@@ -0,0 +1,5 @@
+.table-OverFlow {
+    overflow-y: auto;
```
Let's start using a modified BEM naming system similar to Patternfly's. (let's discuss specifics in chat)
So for now, let's name this --kogito-runtime-tooling--table-container
And, I think we need overflow-x here, not y, so that we ensure a horizontal scroll when needed.

srambach commented on packages/management-console/src/components/Templates/DomainExplorerPage/DomainExplorerPage.css:
```diff
@@ -0,0 +1,5 @@
+    overflow-y: auto;
+}
+
+table.pf-c-table td {min-width: 10em;}
```
This will apply to every PF table on the page, which right now is just one, but this isn't really good practice to rely on. So a better way would be to add another class to that table, say `.kogito-runtime-tooling--domain-explorer__table` and then we can apply this style just to that class.
Also, even though it's just one attribute, I'd put the `min-width:10em;` on its own line.
```
.kogito-runtime-tooling--domain-explorer__table {
  min-width: 10em;
}
```

Sara4994 replied:
done

srambach commented on packages/management-console/src/components/Organisms/DomainExplorerTable/DomainExplorerTable.tsx:
```diff
@@ -0,0 +1,112 @@
+          cells={finalKeys}
+          rows={values}
+          aria-label="Domain Explorer Table"
+        >
```
(See comment on the CSS file)
Add `className="kogito-runtime-tooling--domain-explorer__table"` to define a new class for styling the cells inside.

Sara4994 replied:
done

Sara4994 commented:
> I see a few things:
> 
> * When switching to a different domain and back again, would it be possible to remember what you had selected previously for that domain?
> * When too many columns are selected, the table overflows and there is no horizontal scroll. I'll look at where we want to put the overflow. We might want to wrap the column in one more div with `overflow-x: auto` for that. -----> Fixed
> * After adding many columns and the removing some, I see columns with no headers, and they are not getting enough space.  ----> This is fixed
>   ![image](https://user-images.githubusercontent.com/19825616/74549984-970c4b80-4f1e-11ea-8fcc-04d902ab50eb.png)
> * Here I had removed bookingNumber but it and another field still appeared (without a column header) ------> This 
>   ![image](https://user-images.githubusercontent.com/19825616/74550233-08e49500-4f1f-11ea-9ff1-8cb977a51b24.png)
> * I selected a few fields, applied it, and then unchecked all of them. "Apply Columns" doesn't remove them to go back to the empty state.
> * I'm seeing the removed column data left over even when the column is removed sometimes. Here I added and then removed columns. Also, there used to be 3 rows and suddenly there are 4. It's hard to tell what's going on since most of the fields are identical.
>   ![image](https://user-images.githubusercontent.com/19825616/74551359-08e59480-4f21-11ea-9c15-71cfda8c5b35.png)
>   Sorry, I know this is a lot for one comment. Let me know if you prefer that I put each thing in a different comment going forward.



> @Sara4994 When you uncheck a column, the list changes immediately to the empty state until the button is pressed. Checking an additional column doesn't do this. Can we have the list stay in place until the user clicks "Apply columns" again?

@srambach Actually i see a bug in the table... guess you shared me a screenshot in an older comment, when you deselect the columns in the dropdown and hit apply filter the table data collapses...
this happens only for element "id". When you select and deselect Id, the older data still stays on the table.. this doesnot happen for anyother element in root level... to test this i added some new element at root level, when i select and deselect the new element i dont see such collapse of table...
To overcome this bug, i had to refresh the table content while deselecting the columns... That the reason it appears like empty state comes  up when you deselect columns

Sara4994 commented:
> @Sara4994 When you uncheck a column, the list changes immediately to the empty state until the button is pressed. Checking an additional column doesn't do this. Can we have the list stay in place until the user clicks "Apply columns" again?

@srambach Actually i see a bug in the table... guess you shared me a screenshot in an older comment, when you deselect the columns in the dropdown and hit apply filter the table data collapses...
this happens only for element "id". When you select and deselect Id, the older data still stays on the table.. this doesnot happen for anyother element in root level... to test this i added some new element at root level, when i select and deselect the new element i dont see such collapse of table...
To overcome this bug, i had to refresh the table content while deselecting the columns... That the reason it appears like empty state comes  up when you deselect columns

srambach commented:
Thanks for the CSS changes and bug fixes. I see one more problem - now that the filter can open at narrow widths, I see that the column list is cleared. 
1. Select several fields *including id*
2. Make screen narrow. Open Filter.
Column selection is reset to nothing.
3. Make screen wide. Column picker still has nothing selected. Open it and select a column. Click Apply Columns.
Previous bug is there - columns are out of sync. 
![image](https://user-images.githubusercontent.com/19825616/74749290-ee206200-5237-11ea-812e-1a0afbd943bc.png)

srambach commented on packages/management-console/src/components/Organisms/DomainExplorerTable/DomainExplorerTable.tsx:
```diff
@@ -0,0 +1,121 @@
+    const finalResult = tableContent.map(item => getKeys(item));
+    finalResult.map(result => {
+      keys.push(result.tempKeys);
+      values.push({ cells: result.tempValue, id: Math.random().toString() });
```
Change to 
`      values.push({ cells: result.tempValue, rowKey: Math.random().toString() });
`

Sara4994 replied:
done

srambach commented on packages/management-console/src/components/Organisms/DomainExplorerTable/DomainExplorerTable.tsx:
```diff
@@ -0,0 +1,121 @@
+          className="kogito-management-console--domain-explorer__table"
+        >
+          <TableHeader />
+          <TableBody rowKey={'iqu911'} />
```
Change to
`          <TableBody rowKey="rowKey" />
`

Sara4994 replied:
done

srambach commented:
OK - those last two lines of code to change are because there was some problem where `id` was conflicting with something else.
I'm assuming the `Math.random()` is just for sample data :-) and once there's real data, it will be a real unique value from the data.
Huge thanks to @mturley for his help on this.

srambach commented on packages/management-console/src/components/Organisms/DomainExplorerColumnPicker/DomainExplorerColumnPicker.tsx:
```diff
@@ -0,0 +1,277 @@
+            {getAllChilds(finalResult, 'props')}
+          </Select>
+          <Button variant="primary" onClick={generateQuery}>
+            Apply Columns
```
Sentence case - use `Apply columns` :-D

Sara4994 replied:
done

srambach commented on packages/management-console/src/components/Organisms/DomainExplorerTable/DomainExplorerTable.tsx:
```diff
@@ -0,0 +1,115 @@
+          props: { colSpan: 8 },
+          title: (
+            <Bullseye>
+              <SpinnerComponent spinnerText="Loading Domain Explorer..." />
```
Sentence case `Loading domain explorer`

Sara4994 replied:
done

srambach commented on packages/management-console/src/components/Organisms/DomainExplorerTable/DomainExplorerTable.tsx:
```diff
@@ -0,0 +1,115 @@
+                  No domain data to display
+                </Title>
+                <EmptyStateBody>
+                  Select a Domain from dropdown to see content
```
Sentence case, and "the" 😸   `Select a domain from the dropdown to see content`

Sara4994 replied:
done

srambach commented on packages/management-console/src/components/Organisms/DomainExplorerColumnPicker/DomainExplorerColumnPicker.tsx:
```diff
@@ -0,0 +1,277 @@
+            onSelect={onSelect}
+            selections={selected}
+            isExpanded={isExpanded}
+            placeholderText="Pick Columns"
```
Sentence case "Pick columns"

Sara4994 replied:
done

srambach commented:
![bitmoji](https://render.bitstrips.com/v2/cpanel/46b92545-79d9-4e83-8ea9-8c7b5cc406aa-9830aca1-f58f-4de1-8575-b4f65d4b95a5-v1.png?transparent=1&palette=1&width=246) Just a couple of comments cleaning up the text. Looks good!

Sara4994 commented:
> ![bitmoji](https://camo.githubusercontent.com/fec5990c8af89fc1f8da9b9930ac46fc6d075baa/68747470733a2f2f72656e6465722e6269747374726970732e636f6d2f76322f6370616e656c2f34366239323534352d373964392d346538332d386561392d3863376235636334303661612d39383330616361312d663538662d346465312d383537352d6234663635643462393561352d76312e706e673f7472616e73706172656e743d312670616c657474653d312677696474683d323436) Just a couple of comments cleaning up the text. Looks good!

@srambach done
